### PR TITLE
Cdo apps refactor

### PR DESCRIPTION
**Pull request by wjordan**
Follow-on PR to #6646 and prerequisite for merging the branch.

Extra set of cookbook refactorings needed to get the end-to-end cdo-apps integration test working using the ec2 driver:

```
code-dot-org/cookbooks/cdo-apps$ time KITCHEN_LOCAL_YAML=../.kitchen.ec2.yml bundle exec kitchen verify

[...]

       Command "curl localhost:8080"
         stdout
           should match "<title>Code\\.org\\ \\[adhoc\\]</title>"

       Command "curl localhost:8081"
         stdout
           should match "<title>Anybody\\ can\\ learn\\ \\|\\ Code\\.org</title>"

       Finished in 1.91 seconds (files took 0.27207 seconds to load)
       2 examples, 0 failures

       Finished verifying <default-ubuntu-1404> (0m18.05s).
-----> Kitchen is finished. (16m15.80s)

real    16m16.522s
user    0m3.276s
sys 0m0.448s
```

### CODE REVIEW

File: cookbooks/cdo-apps/libraries/cdo_apps.rb [hunk 1]
```diff
@@ -0,0 +1,96 @@
+module CdoApps
+  def setup_app(app_name)
```
Comment (philbogle) [inline cookbooks/cdo-apps/libraries/cdo_apps.rb:2]: Please document what this helper does.  (It appears to be restricted to pegasus and dashboard due to the specific set of rake tasks and directory structure it requires.)

File: cookbooks/cdo-apps/recipes/crontab.rb [hunk 2]
```diff
@@ -1,17 +1,61 @@
...
+root = File.join node[:home], node.chef_environment
+bin = File.join root, 'bin'
 
-template "/home/#{node[:current_user]}/#{node.chef_environment}/crontab" do
-  source 'crontab.erb'
-  user node[:current_user]
-  group node[:current_user]
-  mode '0644'
-  notifies :run, 'execute[install-crontab]', :immediately
+def cronjob(m:nil, at:nil, cmd:, h:nil, notify:'')
```
Comment (philbogle) [inline cookbooks/cdo-apps/recipes/crontab.rb]: Would it make sense to put the m argument following the h argument?   (I realize these are named arguments so they can be supplied in any order, but the seemingly arbitrary ordering here made it non-obvious what "m" and "h" even stand for.)

File: cookbooks/cdo-apps/recipes/crontab.rb [hunk 3]
```diff
@@ -1,17 +1,61 @@
...
+  end
 end
 
-execute 'install-crontab' do
-  command 'cat ./crontab | crontab -'
-  cwd "/home/#{node[:current_user]}/#{node.chef_environment}"
-  user node[:current_user]
-  group node[:current_user]
-  action :nothing
+# for multi-instance envs (ie production) there should be one daemon,
```
Comment (philbogle) [inline cookbooks/cdo-apps/recipes/crontab.rb]: s/for/For/

File: cookbooks/cdo-apps/recipes/crontab.rb [hunk 4]
```diff
@@ -1,17 +1,61 @@
...
+# so cronjobs that run once per environment go here (standalone env
+# instances are all their own daemon)
+if node['cdo-apps']['daemon']
+  unless node.chef_environment == 'production' # non-production daemons
+    cronjob at: :reboot, cmd: "#{bin}/solr-server > #{root}/pegasus/log/solr.log 2>&1"
```
Comment (philbogle) [inline cookbooks/cdo-apps/recipes/crontab.rb]: Will this ensure that solr is running after initial install?

File: cookbooks/cdo-apps/recipes/crontab.rb [hunk 5]
```diff
@@ -1,17 +1,61 @@
...
+  cronjob m: 35,     cmd: "#{bin}/cron/analyze_hoc_activity"
+  cronjob h: 8, m: 55, cmd: "#{bin}/cron/ops_data_pull"
+  cronjob h: 5, m: 45, cmd: "#{bin}/cron/admin_stats"
+  cronjob h: 4, m: 40, cmd: "#{bin}/cron/funometer"
+
 end
+
+# cronjobs that run on all instances in all environments go here:
+
+# Restart at 0, 4, 8, 12, 16, 20  UTC (4AM/PM, 8AM/PM, 12AM/PM) over the holidays since we won't be deploying
```
Comment (philbogle) [inline cookbooks/cdo-apps/recipes/crontab.rb]: This comment is misleading, could you please update?

File: cookbooks/cdo-repository/libraries/git.rb [hunk 6]
```diff
@@ -0,0 +1,62 @@
+# Patch Chef::Provider::Git to support creating/updating shallow clones.
+class Chef
+  class Provider
+    class Git < Chef::Provider
+
+      def clone
```
Comment (philbogle) [inline cookbooks/cdo-repository/libraries/git.rb:19]: Patching could be fragile when we update chef.  Is it absolutely critical to do this or could we just define a new method that has the desired behavior?

File: cookbooks/cdo-repository/recipes/default.rb [hunk 7]
```diff
@@ -3,44 +3,35 @@
...
 
-include_recipe "cdo-github-access"
-
-template "/home/#{node[:current_user]}/.gemrc" do
-  source 'gemrc.erb'
-  user node[:current_user]
-  group node[:current_user]
+# Sync repo via SSH if key is provided.
+include_recipe 'cdo-github-access'
+has_ssh_key = node['cdo-github-access'] && node['cdo-github-access']['id_rsa'] != ''
```
Comment (philbogle) [inline cookbooks/cdo-repository/recipes/default.rb:8]: What is the desired behavior if node['cdo-github-access']['id_rsa']  is nil/missing?

File: cookbooks/cdo-repository/recipes/default.rb [hunk 8]
```diff
@@ -3,44 +3,35 @@
...
 
-  # Sync the deploy branch to the environment branch.
-  action :checkout
-  user node[:current_user]
-
-  # Set the upstream branch to the appropriate origin.
-  notifies :run, "execute[select-upstream-branch]", :immediately
-end
+  # Default checkout-only for CI-managed instances. (CI script manages pull on updates)
+  action = :checkout
```
Comment (philbogle) [inline cookbooks/cdo-repository/recipes/default.rb]: This code is tricky because of the mutable state in `action`, which can be repeatedly updated by each one of the postfix if statements.  (The non-careful reader might not notice that action = nothing trumps action = sync for an adhoc instance with a shared volume.)

I believe the following variation which gets rid of the mutability is more obvious (also, it avoids having a local and a method with the same name.)

```
# Skip git-repo sync when using a shared volume to prevent data loss on the host.
if GitHelper.shared_volume?(git_path, home_path)
  action :nothing

# Sync instead of checkout for adhoc instances that aren't CI-managed.
elsif node.chef_environment == 'adhoc'
   action :sync

# Default checkout-only for CI-managed instances. (CI script manages pull on updates)
else
  action :checkout
end
```

File: Gemfile [hunk 9]
```diff
@@ -21,6 +21,7 @@ gem 'dalli' # memcached
 gem 'parallel'
 
 gem 'google-api-client'
+gem 'sprockets-derailleur' # Multi-cpu assets precompile
```
Comment (philbogle) [inline Gemfile]: Is sprockets-derailleur an essential part of this change or just an optimization?  If it's non-critical, it might be preferable to factor this into a separate PR, to optimize for getting your key changes in and not having to take a detour to fix any unexpected regressions (e.g. due to concurrency bugs) which might be exist with sprockets-derailleur.

**Comment by philbogle** [comment]
Done with my initial review pass.

File: cookbooks/cdo-repository/recipes/default.rb (hunk 7 above)
Comment (wjordan) [inline cookbooks/cdo-repository/recipes/default.rb:8]: If the key isn't set, the default `node['cdo-repository']['url']` attribute will be used as the git clone url (set to `'https://github.com/code-dot-org/code-dot-org.git'` by default in `attributes.rb`), making it possible to use this cookbook to clone the public github repository without requiring our private credentials within the Chef environment.

I could refactor this for a bit more readability (possibly collapsing cdo-github-access cookbook into cdo-repository, or at least moving the https and ssh attribute-defaults in the same place in the code) if I spend more time on it.

File: cookbooks/cdo-repository/libraries/git.rb (hunk 6 above)
Comment (wjordan) [inline cookbooks/cdo-repository/libraries/git.rb:19]: Not absolutely critical to patch, I just found it quicker to modify the existing Chef resource in this way rather than figure out how to build up a new [custom Chef resource](https://docs.chef.io/custom_resources.html) that reproduces the original behavior and includes the few patches here.

I can figure out how to refactor this into a separate resource extending the original (and it would be helpful to know how to create custom resources anyway).

**Comment by wjordan** [comment]
@philbogle thanks for the review- I'm going to extract the crontab-refactoring portion into a separate PR to merge later since it's not a prerequisite to the rest of this work.

**Comment by wjordan** [comment]
I've extracted the sprockets-derailleur changes from this PR. The asset precompilation optimization is necessary in order to get the integration test to pass without timing out on CircleCI:

```
 adhoc: Precompiling <b>dashboard</b> assets...
              RAILS_ENV=adhoc RACK_ENV=adhoc bundle exec rake assets:precompile

command ./test.sh took more than 10 minutes since last output
```

(Similarly, the build occasionally fails on the git clone step, and I've been working on optimizing that step as well.)

However, these optimizations are not strictly necessary for now since the tests still pass when run locally, so I can return to them in future PRs.

**Comment by philbogle** [comment]
LGTM, thanks.
